### PR TITLE
Fix mongodb entity listeners

### PR DESCRIPTION
**Pull request by rustamwin**
Issues #1527

### CODE REVIEW

**Comment by vlapo** [comment]
@rustamwin Did you tests all hooks? I know this PR is just about fix issue with `afterLoad`. But I would love to see tests to all listeners and subscribers in mongodb, if it is possible.

Btw I do not see tests for listeners and subscribers in full scope. Maybe proposal to new PR.

**Comment by rustamwin** [comment]
@vlapo you are right, this PR is about fix issue with `afterLoad`.

I'll reference new PR for mongodb tests

**Comment by rustamwin** [comment]
@vlapo I think there is all tests about mongodb entity events

File: src/query-builder/transformer/DocumentToEntityTransformer.ts [hunk 1]
```diff
@@ -13,10 +15,12 @@ export class DocumentToEntityTransformer {
     // -------------------------------------------------------------------------
 
     constructor(// private selectionMap: AliasMap,
-                // private joinMappings: JoinMapping[],
-                // private relationCountMetas: RelationCountAttribute[],
-                private enableRelationIdValues: boolean = false
+        // private joinMappings: JoinMapping[],
+        // private relationCountMetas: RelationCountAttribute[],
+        private queryRunner: MongoQueryRunner,
```
Comment (vlapo) [inline src/query-builder/transformer/DocumentToEntityTransformer.ts]: I really do not like pass `QueryRunner` into `DocumentToEntityTransformer` class. This class have only one simple purpose - convert object to entity - so we should keep it simple.

File: src/query-builder/transformer/DocumentToEntityTransformer.ts [hunk 2]
```diff
@@ -177,8 +181,11 @@ export class DocumentToEntityTransformer {
                 }
             });
         });*/
+        const broadcastResult = new BroadcasterResult();
+        this.queryRunner.broadcaster.broadcastLoadEventsForAll(broadcastResult, metadata, [entity]);
+        if (broadcastResult.promises.length > 0) Promise.all(broadcastResult.promises).then(entity => entity);
```
Comment (vlapo) [inline src/query-builder/transformer/DocumentToEntityTransformer.ts]: This is main problem in your PR. You are running async operation like `Promise.all(broadcastResult.promises).then(entity => entity);` inside sync method. I think you are lucky your tests are running successfully. 

Note: I hate one line if statemens 😄

**Comment by vlapo** [comment]
@rustamwin please see comments in code. I have to say this solution is not good. Do not have so much knowledge about listeners/subscribers but check these lines in `SelectQueryBuilder`
https://github.com/typeorm/typeorm/blob/b1d4d7358e60f824725cfd7603e9484e46356436/src/query-builder/SelectQueryBuilder.ts#L1842-L1847
You can see that we have to wait until all promises in `BroadcasterResult` are resolved. So We have to do same in mongodb. 

Luckily `cursor.toArray` and `cursor.next` methods returns `Promise` or call `callback`. So my suggestion is move broadcasting `AfterLoad` event there and make sure all promises are resolved and possible rejections handled.

File: src/query-builder/transformer/DocumentToEntityTransformer.ts (hunk 1 above)
Comment (rustamwin) [reply src/query-builder/transformer/DocumentToEntityTransformer.ts]: But `QueryRunner` pass into other transformers too

Comment (vlapo) [reply src/query-builder/transformer/DocumentToEntityTransformer.ts]: Sure in `RawSqlResultsToEntityTransformer` but it is used to sync create entity not to async operations like broadcasting. This is not the root problem.

File: src/query-builder/transformer/DocumentToEntityTransformer.ts (hunk 2 above)
Comment (rustamwin) [reply src/query-builder/transformer/DocumentToEntityTransformer.ts]: So I'm should run this operation inside `cursor.toArray` ?

Comment (rustamwin) [reply src/query-builder/transformer/DocumentToEntityTransformer.ts]: when run inside `cursor.toArray` worked properly
```js
return ParentCursor.prototype.toArray.call(this).then(async (results: Entity[]) => {
                    const transformer = new DocumentToEntityTransformer();
                    const entities = transformer.transformAll(results, metadata);
                    const broadcastResult = new BroadcasterResult();
                    queryRunner.broadcaster.broadcastLoadEventsForAll(broadcastResult, metadata, entities);
                    if (broadcastResult.promises.length > 0) await Promise.all(broadcastResult.promises);
                    return entities;
                });
```

Comment (vlapo) [reply src/query-builder/transformer/DocumentToEntityTransformer.ts]: Yeah. Whole this `applyEntityTransformationToCursor` is really big hack of mongo 😄 

```typescript
protected applyEntityTransformationToCursor<Entity>(metadata: EntityMetadata, cursor: Cursor<Entity> | AggregationCursor<Entity>) {
        const ParentCursor = PlatformTools.load("mongodb").Cursor;
        const queryRunner = this.queryRunner;
        cursor.toArray = function (callback?: MongoCallback<Entity[]>) {
            if (callback) {
                ParentCursor.prototype.toArray.call(this, (error: MongoError, results: Entity[]): void => {
                    if (error) {
                        callback(error, results);
                        return;
                    }

                    // transform
                    const transformer = new DocumentToEntityTransformer();
                    const entities = transformer.transformAll(results, metadata);

                    // broadcast
                    const broadcastResult = new BroadcasterResult();
                    queryRunner.broadcaster.broadcastLoadEventsForAll(broadcastResult, metadata, entities);

                    Promise.all(broadcastResult.promises).then(() => {
                        return callback(error, entities);
                    });
                });
            } else {
                return ParentCursor.prototype.toArray.call(this).then((results: Entity[]) => {
                    // transform
                    const transformer = new DocumentToEntityTransformer();
                    const entities = transformer.transformAll(results, metadata);

                    // broadcast
                    const broadcastResult = new BroadcasterResult();
                    queryRunner.broadcaster.broadcastLoadEventsForAll(broadcastResult, metadata, entities);

                    return Promise.all(broadcastResult.promises);
                });
            }
        };
        cursor.next = function (callback?: MongoCallback<CursorResult>) {
            if (callback) {
                ParentCursor.prototype.next.call(this, (error: MongoError, result: CursorResult): void => {
                    if (error || !result) {
                        callback(error, result);
                        return;
                    }

                    // transform
                    const transformer = new DocumentToEntityTransformer();
                    const entities = transformer.transform(result, metadata);

                    // broadcast
                    const broadcastResult = new BroadcasterResult();
                    queryRunner.broadcaster.broadcastLoadEventsForAll(broadcastResult, metadata, entities);

                    Promise.all(broadcastResult.promises).then(() => {
                        return callback(error, entities);
                    });
                });
            } else {
                return ParentCursor.prototype.next.call(this).then((result: Entity) => {
                    if (!result) return result;

                    // transform
                    const transformer = new DocumentToEntityTransformer();
                    const entities = transformer.transform(result, metadata);

                    // broadcast
                    const broadcastResult = new BroadcasterResult();
                    queryRunner.broadcaster.broadcastLoadEventsForAll(broadcastResult, metadata, entities);

                    return Promise.all(broadcastResult.promises);
                });
            }
        };
    }
```

**Comment by rustamwin** [comment]
@vlapo Done

**Comment by vlapo** [comment]
Please revert all code style changes not related to this PR.

**Comment by rustamwin** [comment]
Sorry I'm use webstorm :smile:

**Comment by vlapo** [comment]
`Broadcaster.ts` and `DocumentToEntityTransformer.ts` too. Yeah I loved webstorm before visual studio code.

**Comment by rustamwin** [comment]
@vlapo I think about creating cursor builder e.g `SelectQueryBuilder`. What about this?

**Comment by vlapo** [comment]
> `Broadcaster.ts` and `DocumentToEntityTransformer.ts` too. Yeah I loved webstorm before visual studio code.

Still end of file new lines in this documents. After that it can be merged.

> @vlapo I think about creating cursor builder e.g `SelectQueryBuilder`. What about this?

What do you mean in term of "cursor builder"?

**Comment by rustamwin** [comment]
> Still end of file new lines in this documents. After that it can be merged.

Why? `.editorconfig` contains this `insert_final_newline = true` so EOF will be automatically

**Comment by rustamwin** [comment]
> What do you mean in term of "cursor builder"?

class for implements mongo functionality (e.g broadcasting, joining, streaming..)

**Comment by vlapo** [comment]
> > Still end of file new lines in this documents. After that it can be merged.
> 
> Why? `.editorconfig` contains this `insert_final_newline = true` so EOF will be automatically

Sure but you did not change these two files so why to reformat code? Unnecessary files in the list of changes.


> > What do you mean in term of "cursor builder"?
> 
> class for implements mongo functionality (e.g broadcasting, joining, streaming..)

This should be really big step. Maybe discuss this with @pleerock and others in separate issue/PR.

**Comment by rustamwin** [comment]
Okay, now ready for merge :)

**Comment by pleerock** [comment]
Thank you!

**Comment by tien** [comment]
Hi, this issue seems to have re-emerged. AfterInsert, AfterUpdate, BeforeInsert, BeforeUpdate doesn't work, but AfterLoad seems to work fine thou.

**Pull request merged**